### DISCUSSION TITLE
No training loss logged + val metrics logged incorrectly when using lightning as a backend

### DISCUSSION THREAD

**Discussion by NickleDave** [Lightning-AI/pytorch-lightning#18849] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi :wave:
I develop a domain-specific library and recently we switched to using lightning as a backend.
We have realized that our logging is not working correctly though (we're using the default Tensorboard logger).
https://github.com/vocalpy/vak/issues/726

Our calls to `self.log` with the loss in the `training_step` do not get saved in the events file, and all the metrics we log on `validation_step` appear to be incorrect--see first attached image.
![image](https://github.com/Lightning-AI/lightning/assets/11934090/0425be69-70a9-41c3-ae84-297f5b04bce8)

However, if I run the exact same model in a script then everything gets logged correctly (tested with [this notebook](https://github.com/vocalpy/vak/blob/mre-train-loss-logged-in-script-not-library/test-lightning-loss-tweetynet.ipynb))--see second + third attached image.
![image](https://github.com/Lightning-AI/lightning/assets/11934090/a3dec6f1-be3a-4355-a9f9-2e75e7361030)
![image](https://github.com/Lightning-AI/lightning/assets/11934090/cbf6c4ce-e6a3-4161-901e-c1d79abe47bd)

Things that might matter:
- We subclass the lightning module, the class hierarchy for the model I'm looking at is: lightning.LightningModule <- [vak.models.Base](https://github.com/vocalpy/vak/blob/main/src/vak/common/trainer.py) <- [vak.models.FrameClassificationModel](https://github.com/vocalpy/vak/blob/main/src/vak/models/frame_classification_model.py) <- [vak.models.TweetyNet](https://github.com/vocalpy/vak/blob/main/src/vak/models/tweetynet.py). Some of this sub-classing happens [by calling `type` directly to make a new sub-class](https://github.com/vocalpy/vak/blob/main/src/vak/models/decorator.py). I know this system is a bit convoluted--we're trying to make it as easy as possible for someone to declare a model--but I'm not sure why it would break logging specifically? I can tell from the progress bar that training is working as expected and the model converges, even though the logs are incorrect 
- The broken logging happens when we run through our library's CLI, which runs a [`__main__` function](https://github.com/vocalpy/vak/blob/main/src/vak/__main__.py) (i.e., `vak train` calls `vak.__main__.main`). But according to the `Trainer` docstring I would guess this is a good practice? It also means that the DDP strategy doesn't work for us, but that's another issue...
- This is for a model with a non-standard training loop--we train on every possible window from a set of spectrograms, so models are done training before one "epoch". See the trainer we use [here](https://github.com/vocalpy/vak/blob/main/src/vak/common/trainer.py). But I replicated that training loop in the notebook and logging seems to work there.

Any guesses as to why running with our library as the main process messes up lightning's logging?  
My best guess is that it's something related to the sub-classing, since the script above that works correctly removes that sub-classing, and then logging seems to work correctly
